pkg/query-service/auth/auth.go

```diff
@@ -207,65 +209,121 @@ type RegisterRequest struct {
-		if apiErr != nil {
-			zap.S().Debugf("CreateOrg failed, err: %v\n", apiErr.Err)
-			return apiErr
-		}
-		groupName = constants.AdminGroup
-		orgId = org.Id
+	hash, err = passwordHash(req.Password)
+	if err != nil {
+		zap.S().Errorf("failed to generate password hash when registering a user", zap.Error(err))
+		return nil, model.InternalError(model.ErrSignupFailed{})
 	}
 
-	if len(users) > 0 {
-		inv, err := validateInvite(ctx, req)
-		if err != nil {
-			return &model.ApiError{Err: err, Typ: model.ErrorUnauthorized}
-		}
-		org, apiErr := dao.DB().GetOrgByName(ctx, req.OrgName)
-		if apiErr != nil {
-			zap.S().Debugf("GetOrgByName failed, err: %v\n", apiErr.Err)
-			return apiErr
-		}
+	user := &model.User{
+		Id:                 uuid.NewString(),
+		Name:               req.Name,
+		Email:              req.Email,
+		Password:           hash,
+		CreatedAt:          time.Now().Unix(),
+		ProfilePirctureURL: "", // Currently unused
+		GroupId:            group.Id,
+		OrgId:              org.Id,
+	}
 
-		groupName = inv.Role
-		if org != nil {
-			orgId = org.Id
-		}
+	return dao.DB().CreateUser(ctx, user)
+}
+
+// RegisterInvitedUser handles registering a invited user
+func RegisterInvitedUser(ctx context.Context, req *RegisterRequest, nopassword bool) (*model.User, *model.ApiError) {
+
+	if req.InviteToken == "" {
+		return nil, model.BadRequest(fmt.Errorf("invite token is required"))
 	}
 
-	group, apiErr := dao.DB().GetGroupByName(ctx, groupName)
+	if !nopassword && req.Password == "" {
+		return nil, model.BadRequest(model.ErrPasswordRequired{})
+	}
+
+	invite, err := ValidateInvite(ctx, req)
+	if err != nil {
+		zap.S().Errorf("failed to validate invite token", err)
+		return nil, model.BadRequest(model.ErrSignupFailed{})
+	}
+
+	// checking if user email already exists, this is defensive but
+	// required as delete invitation and user creation dont happen
+	// in the same transaction at the end of this function
+	userPayload, apierr := dao.DB().GetUserByEmail(ctx, invite.Email)
+	if apierr != nil {
+		zap.S().Debugf("failed to get user by email", apierr.Err)
+		return nil, apierr
+	}
+
+	if userPayload != nil {
+		// user already exists
+		return &userPayload.User, nil
+	}
+
+	if invite.OrgId == "" {
+		zap.S().Errorf("failed to find org in the invite")
+		return nil, model.InternalError(fmt.Errorf("invalid invite, org not found"))
+	}
+
+	if invite.Role == "" {
+		// if role is not provided, default to viewer
+		invite.Role = constants.ViewerGroup
+	}
+
+	group, apiErr := dao.DB().GetGroupByName(ctx, invite.Role)
 	if apiErr != nil {
 		zap.S().Debugf("GetGroupByName failed, err: %v\n", apiErr.Err)
-		return apiErr
+		return nil, model.InternalError(model.ErrSignupFailed{})
 	}
 
-	hash, err := passwordHash(req.Password)
-	if err != nil {
-		return &model.ApiError{Err: err, Typ: model.ErrorUnauthorized}
+	var hash string
+
+	// check if password is not empty, as for SSO case it can be
+	if req.Password != "" {
```

Comment:
The client doesn't do this but shouldn't we ensure that this branch is executed only when sso is not enabled along with the password requirement? I would consider refactoring this that conveys the SSO context more explicitly.

Reply:
SSO is ee concept. It has no binding in base code. And ideally should not appear in base code anywhere IMO.

Reply:
In ee, there would be OIDC, Oauth (google), and also we are discussing of using passwords from keycloak. 

we are likely to replace auth pkg with an interface and have separate implementation for each of those types. 

I will add this to a separate issue to handle